From 00e5ee6ad83a8f80c8e90af191570dc6cfcd9b40 Mon Sep 17 00:00:00 2001
From: Hin-Tak Leung <htl10@users.sourceforge.net>
Date: Sat, 18 Nov 2023 01:27:42 +0000
Subject: [PATCH] Hook up SkSVGOpenTypeSVGDecoder::Make to enable OT-SVG

"skia/dm/DM.cpp" and "skia/tools/viewer/Viewer.cpp" do this:

    SkGraphics::Init();
    SkGraphics::SetOpenTypeSVGDecoderFactory(SkSVGOpenTypeSVGDecoder::Make);

so that "SkGraphics::GetOpenTypeSVGDecoderFactory()" returns true.

"SetOpenTypeSVGDecoderFactory()" only needs to be called once ever.

Signed-off-by: Hin-Tak Leung <htl10@users.sourceforge.net>
---
 src/ports/SkFontHost_FreeType.cpp | 8 ++++++++
 src/ports/SkScalerContext_win_dw.cpp | 8 ++++++++
 2 file changed, 16 insertions(+)

diff --git a/src/ports/SkFontHost_FreeType.cpp b/src/ports/SkFontHost_FreeType.cpp
index 1f460fe..ee10d75 100644
--- a/src/ports/SkFontHost_FreeType.cpp
+++ b/src/ports/SkFontHost_FreeType.cpp
@@ -60,6 +60,10 @@
 #include <freetype/t1tables.h>
 #include <freetype/ftfntfmt.h>
 
+#if defined(SK_ENABLE_SVG)
+#include "modules/svg/include/SkSVGOpenTypeSVGDecoder.h"
+#endif
+
 using namespace skia_private;
 
 namespace {
@@ -999,6 +1003,10 @@ SkScalerContext_FreeType::SkScalerContext_FreeType(sk_sp<SkTypeface_FreeType> ty
         // FT_LOAD_COLOR with scalable fonts means allow SVG.
         // It also implies attempt to render COLR if available, but this is not used.
 #if defined(FT_CONFIG_OPTION_SVG)
+#if defined(SK_ENABLE_SVG)
+        if (!SkGraphics::GetOpenTypeSVGDecoderFactory())
+            SkGraphics::SetOpenTypeSVGDecoderFactory(SkSVGOpenTypeSVGDecoder::Make);
+#endif
         if (SkGraphics::GetOpenTypeSVGDecoderFactory()) {
             fLoadGlyphFlags |= FT_LOAD_COLOR;
         }
diff --git a/src/ports/SkScalerContext_win_dw.cpp b/src/ports/SkScalerContext_win_dw.cpp
index 68b4744..c99752a 100644
--- a/src/ports/SkScalerContext_win_dw.cpp
+++ b/src/ports/SkScalerContext_win_dw.cpp
@@ -45,6 +45,10 @@
 #include "src/utils/win/SkHRESULT.h"
 #include "src/utils/win/SkTScopedComPtr.h"
 
+#if defined(SK_ENABLE_SVG)
+#include "modules/svg/include/SkSVGOpenTypeSVGDecoder.h"
+#endif
+
 #include <dwrite.h>
 #include <dwrite_1.h>
 #include <dwrite_3.h>
@@ -2275,6 +2279,10 @@ bool SkScalerContext_DW::drawSVGImage(const SkGlyph& glyph, SkCanvas& canvas) {
         return false;
     }
 
+#if defined(SK_ENABLE_SVG)
+    if (!SkGraphics::GetOpenTypeSVGDecoderFactory())
+        SkGraphics::SetOpenTypeSVGDecoderFactory(SkSVGOpenTypeSVGDecoder::Make);
+#endif
     SkGraphics::OpenTypeSVGDecoderFactory svgFactory = SkGraphics::GetOpenTypeSVGDecoderFactory();
     if (!svgFactory) {
         return false;
-- 
2.42.0

